### ISSUE TITLE
zip64 extra data filed parse

### ISSUE THREAD

**Issue by LeeYoung624** [srikanth-lingala/zip4j#68]
As the [issue#65](https://github.com/srikanth-lingala/zip4j/issues/65) has mentioned, zip4j has some parsing errors in zip64 extra field.
The fix in this issue is [here](https://github.com/srikanth-lingala/zip4j/commit/5676742a4ca2e3d8480d966f165100d02798363f), but I think there are some problems with this fix.

The zip specification said about extra data :
> 4.5.3 -Zip64 Extended Information Extra Field (0x0001):
      The following is the layout of the zip64 extended information "extra" block. If one of the size or offset fields in the Local or Central directory record is too small to hold the required data, a Zip64 extended information record is created. The order of the fields in the zip64 extended information record is fixed, but the fields MUST only appear if the corresponding Local or Central directory record field is set to 0xFFFF or 0xFFFFFFFF.

As the specification said, the extra field should only be valid when the corresponding original field is 0xFFFF or 0xFFFFFFFF. The fix in commit [Do not override from Zip64 record if value not present](https://github.com/srikanth-lingala/zip4j/commit/5676742a4ca2e3d8480d966f165100d02798363f) is here:
```
    if (zip64ExtendedInfo.getUncompressedSize() != -1) {
      fileHeader.setUncompressedSize(zip64ExtendedInfo.getUncompressedSize());
    }

    if (zip64ExtendedInfo.getCompressedSize() != -1) {
      fileHeader.setCompressedSize(zip64ExtendedInfo.getCompressedSize());
    }

    if (zip64ExtendedInfo.getOffsetLocalHeader() != -1) {
      fileHeader.setOffsetLocalHeader(zip64ExtendedInfo.getOffsetLocalHeader());
    }

    if (zip64ExtendedInfo.getDiskNumberStart() != -1) {
      fileHeader.setDiskNumberStart(zip64ExtendedInfo.getDiskNumberStart());
    }
```

I think it should be like this:
```
    if (fileHeader.getUncompressedSize() != 0xFFFFFFFFL) {
      fileHeader.setUncompressedSize(zip64ExtendedInfo.getUncompressedSize());
    }

    if (fileHeader.getCompressedSize() != 0xFFFFFFFFL) {
      fileHeader.setCompressedSize(zip64ExtendedInfo.getCompressedSize());
    }

    if (fileHeader.getOffsetLocalHeader() != 0xFFFFFFFFL) {
      fileHeader.setOffsetLocalHeader(zip64ExtendedInfo.getOffsetLocalHeader());
    }

    if (fileHeader.getDiskNumberStart() != 0xFFFFL) {
      fileHeader.setDiskNumberStart(zip64ExtendedInfo.getDiskNumberStart());
    }
```

**Comment by srikanth-lingala** [comment]
Those checks you mentioned are not related to the point you mentioned from the documentation. What the documentation is saying is that if any of the header value (size or crc) is set to 0xFFFFFFFF, then we can expect the `Zip64 Extended Information Extra Field` to be present as part of the headers. That is the whole purpose of that zip64 field. If any of the value cannot be stored in 4 bytes, then it has to be stored in 8 byte fields which are part of the zip64 header fields (0xFFFFFFFF is the maximum value that can be written in 4 bytes).

What these checks as part of fix for #65 are for is: in the past some users reported some zip files where not all fields were present in this zip64 header. Some tools created zip files with zip64 headers with only part of the data. This check is to override the data only if the data is present, else skip it. Hope this makes sense.

**Comment by LeeYoung624** [comment]
I understand it now. It's really impressing that you are so familiar with zip specification.
Thank you very much.
